applications/ehr_query_llm/lmm/run_lmm.sh

```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
+
+# Array to hold PIDs of background processes
+declare -a bg_pids
+
+set -x 
+
+# Function to clean up and kill the background processes
+cleanup() {
+    echo "Terminating Llama.cpp server process..."
+    # Kill the /bin/server process.
+    pkill -f "/opt/nvidia/holoscan/llama.cpp/build/bin/server"
+
+    # Kill all background Python processes
+    for pid in "${bg_pids[@]}"; do
+        if kill -0 $pid > /dev/null 2>&1; then
+            kill $pid
+        fi
+    done
+
+    exit 0
+}
+
+set_python_path() {
+    # Set the python path to the holoscan_sdk_install path
+    local holoscan_sdk_install=$(grep -Po '^holoscan_DIR:PATH=\K[^ ]*' build/CMakeCache.txt)
+    local holohub_build_dir="${SCRIPT_DIR}/build"
+    local environment="export PYTHONPATH=${holoscan_sdk_install}/../../../python/lib:${holohub_build_dir}/python/lib:${SCRIPT_DIR}"
+    echo "${environment}"
+    eval "${environment}"
+}
+
+set_transformer_cache() {
+    local transformer_cache="export HF_HOME=/workspace/volumes/models"
+    echo "${transformer_cache}"
+    eval "${transformer_cache}"
+}
+
+set_offline_flags() {
+    local offline_flags="export TRANSFORMERS_OFFLINE=1 ANONYMIZED_TELEMETRY=False"
+    echo "${offline_flags}"
+    eval "${offline_flags}"
+}
+
+# Trap SIGINT (Ctrl-C) and SIGTERM signals and call the cleanup function
+trap cleanup SIGINT SIGTERM
+
+# Set environment variables
+set_python_path
+set_transformer_cache
+# set_offline_flags
+netstat -tuln | grep -E ':8080|:8050|:49000'
+
+# Build ehr_query_llm/lmm
+/workspace/holohub/run build ehr_query_llm/lmm
+
+# Define log file paths
+LLM_LOG_FILE="/workspace/holohub/applications/ehr_query_llm/lmm/llama_cpp.log"
+
+echo "Hold on tight, starting the Llama.cpp LLM server process + main EHR app!"
+
+# Run the Llama.cpp LLM server process + main EHR app
+/opt/nvidia/holoscan/llama.cpp/build/bin/server \
+    -m  /workspace/volumes/models/openchat-3.5-0106.Q8_0.gguf \
+    --host 0.0.0.0 \
+    --port 8080 \
+    -ngl 1000 \
+    -c 8000 \
+    --cont-batching \
+    > ${LLM_LOG_FILE} 2>&1 \
+& python3 /workspace/holohub/applications/ehr_query_llm/lmm/asr_llm_tts.py &
+bg_pids+=($!) 
+
+sleep 5  # Give the server time to start
+if ! curl -s -o /dev/null -w "%{http_code}" http://localhost:8080 | grep -q "200"; then
```

Comment:
@cdinea this part you added is mistaken; the command returns 0 on success, so this gives false positives when Llama.cpp has started successfully. Unfortunately, or fortunately, the `exit` in the conditional block did not work, so the script carries on. Please correct this in your next commit, and ensure to sign off the commit with `git commit -s`

Reply:
no, I believe exit 1 is ok to be used here .
In Unix/Linux systems, exit codes follow these conventions:

- exit 0 indicates successful execution
- exit 1 (or any non-zero value) indicates an error or failure

In this case, I am checking if the server failed to start. If the check fails, that's an error condition, so returning 1 is the correct behavior

Reply:
I recommend taking a look [here](https://www.agileconnection.com/article/overview-linux-exit-codes) for further reference

Reply:
@cdinea before dismissing it, please
1. check line 75 more carefully
3. run the script and validate, specifically looking for the output stating server failed.

Reply:
I did check it and the script is written correctly.Do you have trouble understanding the if statement? should I break it down piece by piece?

Reply:
Here is a breakdown for easier understanding 
curl command components:
-s: Silent mode (doesn't show progress bar or error messages)
-o /dev/null: Discards the actual response body by sending it to /dev/null
-w "%{http_code}": Only outputs the HTTP status code of the response
http://localhost:8080/: The URL of the Llama.cpp server
The command pipeline:
The curl output (just the status code) is piped to grep -q "200"
grep -q "200" silently checks if the status code is 200 (HTTP OK)
The -q flag makes grep quiet (no output, just return status)
The logic:
The ! at the start inverts the condition
If the grep doesn't find "200" (meaning the server isn't responding correctly):
It prints the error message
Exits with status code 1 (indicating failure)
In plain English, this code is saying: "Try to contact the server at localhost:8080. If we don't get back a successful HTTP 200 response, print an error message and exit the script with a failure status."

Reply:
My terminal window is large but not large enough to retain the output way back at the beginning of the tests. False positive was reported even though everything carried on, and there is no retry in testing the Web server response after sleeping 5s.

Reply:
it is hard to understand the issue if there is no specific log .Also, I am not sure what "way back at the beginning of tests mean" ? did you only test this once when I first uploaded the code? - the if statement logic that you specifically added the comment for is correct so, if the issues that you reported were found , it might be because of prior logic : the server might need more than 5 secs to load the LLM into memory, initialize GPU resources and start listening on the port . However, the error you reported is not directly linked to this line based on your description and therefore I can only try to repro on my end next week to confirm the "false positive" occurrences.

Reply:
I actually reproduced the false positive, and in closer examination, only the unexpected "server failed...` log entry was echo'ed, but `exit` was not called. It is puzzling. See below for the output. I can give you the screenshot too.

Also, `netstat` is not installed in the dev container, hence the failure when the shell script calls, but the script is not set to fail when any commands fails. From the look of it, the netstat is for info only 

```
++ HF_HOME=/workspace/volumes/models
+ netstat -tuln
+ grep -E ':8080|:8050|:49000'
./applications/ehr_query_llm/lmm/run_lmm.sh: line 52: netstat: command not found
+ /workspace/holohub/run build ehr_query_llm/lmm
Building Holohub
Building ehr_query_llm/lmm application
[command] cmake -S . -B build/ehr_query_llm/lmm --no-warn-unused-cli -DPython3_EXECUTABLE=/usr/bin/python3 -DPython3_ROOT_DIR=/usr/lib/python3 -DHOLOHUB_DATA_DIR:PATH=/workspace/holohub/data -DCMAKE_BUILD_TYPE=release -DAPP_ehr_query_llm/lmm=1
Not searching for unused variables given on the command line.
-- Configuring done
-- Generating done
-- Build files have been written to: /workspace/holohub/build/ehr_query_llm/lmm
[command] cmake --build build/ehr_query_llm/lmm -j
Holohub build done.
+ LLM_LOG_FILE=/workspace/holohub/applications/ehr_query_llm/lmm/llama_cpp.log
+ echo 'Hold on tight, starting the Llama.cpp LLM server process + main EHR app!'
Hold on tight, starting the Llama.cpp LLM server process + main EHR app!
+ bg_pids+=($!)
+ sleep 5
+ python3 /workspace/holohub/applications/ehr_query_llm/lmm/asr_llm_tts.py
+ /opt/nvidia/holoscan/llama.cpp/build/bin/server -m /workspace/volumes/models/openchat-3.5-0106.Q8_0.gguf --host 0.0.0.0 --port 8080 -ngl 1000 -c 8000 --cont-batching
+ curl -s -o /dev/null -w '%{http_code}' http://localhost:8080
+ grep -q 200
+ echo 'Llama.cpp server failed to start or respond'
Llama.cpp server failed to start or respond
+ exit 1
root@mingq-dt:~# ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm_route.c:877:(find_matching_chmap) Found no matching channel map
DEBUG - urllib3.connectionpool: Starting new HTTP connection (1): 127.0.0.1:8080
Attempting to connect to LLM server...
DEBUG - urllib3.connectionpool: Starting new HTTP connection (1): 127.0.0.1:8080
Attempting to connect to LLM server...
DEBUG - urllib3.connectionpool: Starting new HTTP connection (1): 127.0.0.1:8080
DEBUG - urllib3.connectionpool: http://127.0.0.1:8080 "GET / HTTP/11" 200 33456
DEBUG - agents.base_agent.SelectorAgent: SelectorAgent connected to Llama.cpp server at http://127.0.0.1:8080
DEBUG - urllib3.connectionpool: Starting new HTTP connection (1): 127.0.0.1:8080
DEBUG - urllib3.connectionpool: http://127.0.0.1:8080 "GET / HTTP/11" 200 [33456]
```

Reply:
thanks for providing the logs - I just looked at the logs and the server and Python script were started with & so that explains why the server started successfully without exiting

I can try to repro this next week and think of ways to make it more robust if the model takes too much time to load.

Reply:
Resolving this thread as a ticket has been created to investigate the observation.